src/ir/principal-type.cpp

```diff
@@ -0,0 +1,1033 @@
+/*
+ * Copyright 2025 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <iostream>
+
+#include "principal-type.h"
+#include "support/disjoint_sets.h"
+
+namespace wasm {
+
+namespace {
+
+#ifndef NDEBUG
+
+bool valid(const VarAbsHeapType& type) {
+  if (!type.ht.isBasic()) {
+    return false;
+  }
+  // Sharedness is represented in `share`, not `ht`.
+  if (type.ht.isShared()) {
+    return false;
+  }
+  return true;
+}
+
+bool valid(const VarDefHeapType& type) { return !type.ht.isBasic(); }
+
+bool valid(const VarHeapType& type) {
+  if (auto* t = std::get_if<VarAbsHeapType>(&type)) {
+    return valid(*t);
+  }
+  if (auto* t = std::get_if<VarDefHeapType>(&type)) {
+    return valid(*t);
+  }
+  return true;
+}
+
+bool valid(const VarRef& type) { return valid(type.ht); }
+
+bool valid(const VarType& type) {
+  if (auto* ref = std::get_if<VarRef>(&type)) {
+    return valid(*ref);
+  }
+  if (auto* t = std::get_if<Type>(&type)) {
+    return t->isSingle() || *t == Type::unreachable;
+  }
+  return true;
+}
+
+#endif // NDEBUG
+
+VarType canonicalizeRef(const VarRef& ref) {
+  const auto* null = std::get_if<Nullability>(&ref.null);
+  if (!null) {
+    return ref;
+  }
```

Comment:
These helpers might benefit from comments, as I feel I'm missing the point. What sense of "canonicalization" is meant here? Is the canonical representation defined somewhere?

Reply:
I've added comments for all the helpers. It might be helpful to start reading the implementation of `PrincipalType::compose` at the end of the file to get a sense for all the helpers are used, too.